toolchain/sem_ir/id_kind.h

```diff
@@ -115,16 +116,19 @@ class TypeEnum {
 
 // An enum of all the ID types used as instruction operands.
 using IdKind = TypeEnum<
-    // From sem_ir/builtin_inst_kind.h.
-    BuiltinInstKind,
     // From base/value_store.h.
     IntId, RealId, FloatId, StringLiteralValueId,
     // From sem_ir/id.h.
     InstId, AbsoluteInstId, MatchingInstId, ConstantId, EntityNameId,
     CompileTimeBindIndex, RuntimeParamIndex, FunctionId, ClassId, InterfaceId,
     ImplId, GenericId, SpecificId, ImportIRId, ImportIRInstId, LocId, BoolValue,
     IntKind, NameId, NameScopeId, InstBlockId, TypeId, TypeBlockId,
-    ElementIndex, LibraryNameId, FloatKind>;
+    ElementIndex, LibraryNameId, FloatKind,
+// Subtyped InstId variants.
+#define CARBON_SEM_IR_INST_KIND(Name) Name##InstId,
+#include "toolchain/sem_ir/inst_kind.def"
```

Comment:
I'm concerned by this part: right now we have a few places where we're uniformly processing `InstId` operands of instructions by looking at the `IdKind`, and after this change those places will no longer visit the `storage_id` field on `Temporary` ([example](https://github.com/carbon-language/carbon-lang/blob/85f6bf32b5a42fcd5d0ac5faa4e4780a68787710/toolchain/check/subst.cpp#L113), [example](https://github.com/carbon-language/carbon-lang/blob/85f6bf32b5a42fcd5d0ac5faa4e4780a68787710/toolchain/check/deduce.cpp#L105)). Maybe we could map all the typed `InstId` kinds to `IdKind::For<InstId>`?

Reply:
PS: I was concerned that collapsing to `IdKind::For<InstId>` would affect the ability of `NodeStack` (and possibly others) to propagate a `TypedInstId`. This is what led the discussion below, where there's maybe not actually enough value, closing the PR.